### PR TITLE
Getting syncdbAfterWrite working

### PR DESCRIPTION
**Pull request by zachmprince**
## What is the change?

This PR modifies `Database3.syncToSharedFolder` so that it closes the `h5py.File` before trying to copy the h5 file to the working directory. The full process is:

1. Call `h5py.File.close()` and set the object to `None`.
2. Copy the h5py file to the working directory.
3. Garbage collect to avoid multiple database instances in memory.
4. Reload the h5 file in append mode

Tests were also added to ensure the `syncDbAfterWrite` setting behaves as expected.

## Why is the change being made?

Partial resolution to #1794

Previously, setting `syncDbAfterWrite` caused a `PermissionError` due to the h5 file not being closed before trying to copy it.

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/bookkeeping/db/database3.py [hunk 1]
```diff
@@ -52,6 +52,7 @@
     Tuple,
     Type,
 )
+import gc
```
Comment (john-science) [inline armi/bookkeeping/db/database3.py:55]: Can you move this up to between lines 35 and 36, so the std lib imports are still in alphabetical order?

Thanks!

File: armi/bookkeeping/db/tests/test_databaseInterface.py [hunk 2]
```diff
@@ -178,6 +178,54 @@ def test_timeNodeLoop_tightCoupling(self):
...
+
+        self.o.cs["syncDbAfterWrite"] = True
+        self.o.cs["burnSteps"] = 2  # make test insensitive to burn steps
+
+        self.dbi.interactBOL()
+        self.assertFalse(os.path.exists(self.dbi.database.fileName))
+
+        # Go through a few time nodes to ensure appending is working
+        for timeNode in range(self.o.cs["burnSteps"]):
+            r.p.cycle, r.p.timeNode = 0, timeNode
```
Comment (john-science) [inline armi/bookkeeping/db/tests/test_databaseInterface.py:196]:
```suggestion
            r.p.cycle = 0
            r.p.timeNode = timeNode
```

Comment (john-science) [reply armi/bookkeeping/db/tests/test_databaseInterface.py:196]: I just prefer clarity.

**Comment by john-science** [comment]
@zachmprince Remind me, did we run those downstream tests successfully?  Last I heard, you sounded convinced, but were worried about some unrelated errors?  How are you feeling now?

**Comment by john-science** [comment]
@zachmprince Remind me, did we run those downstream tests successfully? Last I heard, you sounded convinced, but were worried about some unrelated errors? How are you feeling now?

**Comment by zachmprince** [comment]
I am confident these changes are not affecting downstream applications. Test failure I ran into is unrelated. If you are okay merging it, please do! It's all okay on my side.

**Review by john-science — APPROVED** [review]

File: armi/bookkeeping/db/tests/test_databaseInterface.py [hunk 3]
```diff
@@ -194,6 +194,55 @@ def test_timeNodeLoop_tightCoupling(self):
...
+                self.assertTrue(self.dbi.database.hasTimeStep(r.p.cycle, tn))
+
+        # Make sure EOL runs smoothly
+        self.dbi.interactEOL()
+        self.assertTrue(os.path.exists(self.dbi.database.fileName))
+
+    def test_noSyncDbAfterWrite(self):
+        """
+        Test to ensure that the fast-path database is NOT copied to working
+        directory at every time node when ``syncDbAfterWrite`` is ``True``.
```
Comment (john-science) [inline armi/bookkeeping/db/tests/test_databaseInterface.py:235]:
```suggestion
        directory at every time node when ``syncDbAfterWrite`` is ``False``.
```

**Pull request merged**